Google.Api.Gax.Grpc/ApiMetadata.cs

```diff
@@ -54,22 +62,46 @@ private ApiMetadata(string name)
         /// </remarks>
         /// <param name="name">The name of the API. Must not be null or empty.</param>
         /// <param name="descriptors">The protobuf descriptors of the API. Must not be null.</param>
-        public ApiMetadata(string name, IEnumerable<FileDescriptor> descriptors) : this(name)
+        public ApiMetadata(string name, IEnumerable<FileDescriptor> descriptors) : this(name, BuildDescriptorsProviderFromDescriptors(descriptors), false)
         {
-            var actualDescriptors = descriptors.ToList().AsReadOnly();
-            _fileDescriptorsProvider = new Lazy<IReadOnlyList<FileDescriptor>>(() => actualDescriptors);
         }
 
+        /// <summary>
+        /// Method used in the above constructor to create the lazy provider.
+        /// (Unfortunately this can't be a local method.)
+        /// </summary>
+        private static Lazy<IReadOnlyList<FileDescriptor>> BuildDescriptorsProviderFromDescriptors(IEnumerable<FileDescriptor> descriptors)
+        {
+            var actualDescriptors = descriptors.ToList().AsReadOnly();
+            return new Lazy<IReadOnlyList<FileDescriptor>>(() => actualDescriptors);
```

Comment:
Shouldn't this be as follows for it to really be lazy? (Can't suggest here)

```csharp
return new Lazy<IReadOnlyList<FieldDescriptor>>(() => descriptors.ToList().AsReadOnly());
```

Reply:
And also, I think you can just put this line of code when calling the other constructor above and remove the static method.

Reply:
The intention isn't for it be lazy in this case - indeed, we want to eagerly evaluate the argument so that any later changes aren't reflected. But we're wrapping it in a Lazy for consistency with the other overload.